mlflow/dspy/save.py

```diff
@@ -0,0 +1,294 @@
+    """
+    Returns:
+        The default Conda environment for MLflow Models produced by calls to `save_model()` and
+        `log_model()`.
+    """
+    return _mlflow_conda_env(additional_pip_deps=get_default_pip_requirements())
+
+
+@experimental
+@format_docstring(LOG_MODEL_PARAM_DOCS.format(package_name=FLAVOR_NAME))
+def save_model(
+    model,
+    path,
+    conda_env=None,
+    mlflow_model=None,
+    signature: ModelSignature = None,
+    input_example: ModelInputExample = None,
+    pip_requirements=None,
+    extra_pip_requirements=None,
+    save_model_kwargs=None,
+    metadata=None,
+):
+    """
+    Save a Dspy model.
+
+    This method saves a Dspy model along with metadata such as model signature and conda
+    environments to local file system. This method is called inside `mlflow.dspy.log_model()`.
+
+    Args:
+        model: an instance of `dspy.Module`. The Dspy model/module to be saved.
+        path: local path where the MLflow model is to be saved.
+        conda_env: {{ conda_env }}
+        mlflow_model: an instance of `mlflow.models.Model`, defaults to None. MLflow model
+            configuration to which to add the Dspy model metadata. If None, a blank instance will
+            be created.
+        signature: {{ signature }}
+        input_example: {{ input_example }}
+        pip_requirements: {{ pip_requirements }}
+        extra_pip_requirements: {{ extra_pip_requirements }}
+        save_model_kwargs: A dict of kwargs to pass save method.
+        metadata: {{ metadata }}
+    """
+
+    import dspy
+
+    if signature is None:
+        _logger.warning("You are saving a dspy model without specifying model signature.")
+    else:
+        num_inputs = len(signature.inputs.inputs)
+        if num_inputs == 0:
+            raise MlflowException(
+                "The model signature's input schema must contain at least one field.",
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+
+    if mlflow_model is None:
+        mlflow_model = Model()
+    if signature is not None:
+        mlflow_model.signature = signature
+    if input_example is not None:
+        path = os.path.abspath(path)
+        _validate_and_prepare_target_save_path(path)
+        _save_example(mlflow_model, input_example, path)
+    if metadata is not None:
+        mlflow_model.metadata = metadata
+
+    save_model_kwargs = save_model_kwargs or {}
+
+    model_data_subpath = _MODEL_DATA_PATH
+    # Construct new data folder in existing path.
+    data_path = os.path.join(path, model_data_subpath)
+    os.makedirs(data_path, exist_ok=True)
+    # Set the model path to end with ".pkl" as we use cloudpickle for serialization.
+    model_subpath = os.path.join(model_data_subpath, _MODEL_SAVE_PATH) + ".pkl"
+    model_path = os.path.join(path, model_subpath)
+    # Dspy has a global context `dspy.settings`, and we need to save it along with the model.
+    dspy_settings = dict(dspy.settings.config)
```

Comment:
Does this save the actual configuration, or is it an object memory reference pointer?

Reply:
good question - it saves the actual config, I added a test case for that.